[expo-sharing] Add expo-sharing module with basic functionality

enterteg opened the pull request:
## Why
Package was response for users request about possibility to share to other platforms, especially instagram

https://expo.canny.io/feature-requests/p/make-it-possible-to-share-an-image-to-instagram-on-ios-and-android

## How
Created new `expo-sharing` package with basic sharing functionality.
On Android used Intent `Intent.ACTION_SEND` and to ensure file is accessible for third applications used `FileProvider`.

On iOS used `UIDocumentInteractionController presentOpenInMenuFromRect` to open action sheet.

For android added `mimeType` property, for iOS added `UTI` to be able to control shown applications that can handle file.

## Test Plan
Test using NCL Sharing module.

enterteg committed:
[expo-sharing] created expo-sharing module

enterteg committed:
[expo-sharing] updated readme

ide commented on packages/expo-sharing/android/src/main/java/expo/modules/sharing/SharingModule.java:
```diff
@@ -0,0 +1,88 @@
+    String path = uri.getPath();
+
+    if (path == null) {
+      promise.reject(new Error(("Url is not valid")));
```
```suggestion
      promise.reject("ERR_SHARING_INVALID_URI", "The URI to the file to share is invalid");
```

ide commented on packages/expo-sharing/android/src/main/java/expo/modules/sharing/SharingModule.java:
```diff
@@ -0,0 +1,88 @@
+    boolean localFile = "file".equals(uri.getScheme());
+
+    if (!localFile) {
+      promise.reject(new Error(("Only local files are supported")));
```
```suggestion
      promise.reject("ERR_SHARING_INVALID_URI", "Only local files are supported");
```

ide commented on packages/expo-sharing/android/src/main/java/expo/modules/sharing/SharingModule.java:
```diff
@@ -0,0 +1,88 @@
+    try {
+      uri = FileProvider.getUriForFile(mContext, mContext.getApplicationInfo().packageName + ".provider", file);
+    } catch (Exception e) {
+      promise.reject(e);
```
```suggestion
      promise.reject("ERR_SHARING", "Failed to get the content URI for the file to share", e);
```

ide reviewed (changes requested):
We also need support for Web Share: https://wicg.github.io/web-share/ and also update the docs.

esamelson commented on packages/expo-sharing/android/src/main/java/expo/modules/sharing/SharingModule.java:
```diff
@@ -0,0 +1,88 @@
+  }
+
+  @ExpoMethod
+  public void shareAsync(String url, ReadableArguments params, final Promise promise) {
```
This is pretty much a special case of the logic in the `expo-intent-launcher` module. Have you thought about just implementing the Android side of this module as a JS wrapper around `IntentLauncher`/`FS` instead? Might need to add another FS method for getting a shareable URL from `FileProvider`, but I wonder if that approach might be better than maintaining essentially the same native logic in two different modules.

tsapeta replied:
Yes, we were thinking of using `IntentLauncher`, but my assumption was that `Sharing` module is for more specific use cases and depending on `IntentLauncher` just for a few lines of code may not make sense since it would add another dependency that the user has to install.

esamelson replied:
Hmm, that makes sense. (It would probably also need to depend on FS => another dependency.) Just throwing this out there -- might it make sense move the more general logic from the `intent-launcher` module into here? It would be nice if we could figure out a way not to need to maintain duplicate native code. Not the end of the world if not, though.

enterteg committed:
expo-sharing fixes after review

sjchmiela committed:
[expo-sharing] Some changes

- add `.gitignore` to `android`
- resolve the promise only after sharing is completed
- add web support
- add `isAvailableAsync` method
- do not allow calling shareAsync multiple times at once

sjchmiela committed:
[ncl] Minor Sharing screen upgrade

sjchmiela committed:
[expo-sharing] Update Markdown content

sjchmiela committed:
[expo-sharing] Add dependency requirements to app and expoview
